rllib/utils/explorations/epsilon_greedy.py

```diff
@@ -0,0 +1,191 @@
+import gym
+import numpy as np
+
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.explorations.exploration import Exploration
+from ray.rllib.utils.framework import try_import_tf, try_import_torch, \
+    get_variable
+from ray.rllib.utils.schedules import ConstantSchedule, PiecewiseSchedule
+
+tf = try_import_tf()
+torch, _ = try_import_torch()
+
+
+class EpsilonGreedy(Exploration):
+    """
+    An epsilon-greedy Exploration class that produces exploration actions
+    when given a Model's output and a current epsilon value (based on some
+    Schedule).
+    """
+
+    def __init__(self,
+                 action_space,
+                 initial_epsilon=1.0,
+                 final_epsilon=0.1,
+                 epsilon_timesteps=int(1e5),
+                 raw_model_output_key="q_values",
+                 fixed_per_worker_epsilon=False,
+                 worker_info=None,
+                 framework="tf"):
+        """
+        Args:
+            action_space (Space): The gym action space used by the environment.
+            initial_epsilon (float): The initial epsilon value to use.
+            final_epsilon (float): The final epsilon value to use.
+            epsilon_timesteps (int): The time step after which epsilon should
+                always be `final_epsilon`.
+            framework (Optional[str]): One of None, "tf", "torch".
+        """
+        # For now, require Discrete action space (may loosen this restriction
+        # in the future).
+        assert isinstance(action_space, gym.spaces.Discrete)
+        assert framework is not None
+        super().__init__(
+            action_space=action_space,
+            worker_info=worker_info,
+            framework=framework)
+
+        self.epsilon_schedule = None
+        # Use a fixed, different epsilon per worker. See: Ape-X paper.
+        if fixed_per_worker_epsilon is True:
```

Comment:
should this be a separate class (e.g., PerWorkerEpsilonGreedy)?

Reply:
Good idea. Will change.

Reply:
done.